includes/shell/ast.hpp

```diff
@@ -38,19 +54,36 @@ class Ast {
 		unique_ptr<Node> right;
 	};
 
+	struct Redirection {
+	public:
+		Redirection() = default;
+		Redirection(Redirection &&other);
+		~Redirection();
+
+	public:
+		void print(int level) const;
+
+	public:
+		string file_name;
+		Token::Type redirection_type;
+		size_t io_number;
```

Comment:
Should it be an `int` since it's a file descriptor?

Reply:
https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_07
this links to an RFC on file descriptors, and that says unsigned so I went with that

Reply:
https://stackoverflow.com/questions/17788529/why-are-unix-file-descriptors-not-implemented-by-their-own-type-especially-in-c

I suppose it's not too bad to have an unsigned type. Most functions that take a file descriptor will take an integer though, so this means we'll have to cast it a lot. Maybe it's worth making a `unique_fd` wrapper, which would throw an error on negative values or something like that.

Reply:
I use `stoul` so that will ensure it's not signed once that passes

Reply:
I think if we decide to represent the fd as unsigned int here, then we should represent it everywhere else as unsigned too for consistency. I'm wondering whether that choice makes sense

Reply:
made it an int